server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java

```diff
@@ -216,39 +242,28 @@ public ClusterFormationState(
         private static boolean calculateHasDiscoveredQuorum(
             List<DiscoveryNode> foundPeers,
             ElectionStrategy electionStrategy,
-            DiscoveryNode localNode,
-            long currentTerm,
-            long acceptedTerm,
-            long clusterStateVersion,
-            VotingConfiguration lastCommittedConfiguration,
-            VotingConfiguration lastAcceptedConfiguration
+            ClusterFormationClusterStateView clusterFormationClusterStateView
         ) {
             final VoteCollection voteCollection = new VoteCollection();
             foundPeers.forEach(voteCollection::addVote);
             return electionStrategy.isElectionQuorum(
-                localNode,
-                currentTerm,
-                acceptedTerm,
-                clusterStateVersion,
-                lastCommittedConfiguration,
-                lastAcceptedConfiguration,
+                clusterFormationClusterStateView.localNode,
+                clusterFormationClusterStateView.currentTerm,
+                clusterFormationClusterStateView.lastAcceptedTerm,
+                clusterFormationClusterStateView.lastAcceptedVersion,
+                clusterFormationClusterStateView.lastCommittedConfiguration,
+                clusterFormationClusterStateView.lastAcceptedConfiguration,
                 voteCollection
             );
         }
 
         public ClusterFormationState(StreamInput in) throws IOException {
             this(
                 in.readStringCollectionAsList(),
-                new DiscoveryNode(in),
-                in.readMap(DiscoveryNode::new),
-                in.readLong(),
-                in.readLong(),
-                new VotingConfiguration(in),
-                new VotingConfiguration(in),
+                new ClusterFormationClusterStateView(in),
                 in.readCollectionAsImmutableList(TransportAddress::new),
                 in.readCollectionAsImmutableList(DiscoveryNode::new),
                 in.readCollectionAsImmutableSet(DiscoveryNode::new),
-                in.readLong(),
```

Comment:
Nice idea, but this reorders the fields as sent on the wire so it won't work.

If we really want to do this we'd need to introduce a new `TransportVersion` for the new order, and write an adapter to shuffle the fields around when reading from an older node.